imdict.js

```diff
@@ -1,167 +1,70 @@
-    return TypedMap.__coerce(Immutable.Map.prototype.merge.apply(this, args));
-  };
-
-  TypedMap.prototype.mergeWith = function (merger) {
-    const args = this.__forceIterables(Array.prototype.slice.call(arguments, 1));
-    args.unshift(merger);
-    return TypedMap.__coerce(Immutable.Map.prototype.mergeWith.apply(this, args));
-  };
-
-  TypedMap.prototype.mergeDeep = function () {
-    const args = this.__forceIterables(Array.prototype.slice.call(arguments));
-    return TypedMap.__coerce(Immutable.Map.prototype.mergeDeep.apply(this, args));
-  };
-
-  TypedMap.prototype.mergeDeepWith = function (merger) {
-    const args = this.__forceIterables(Array.prototype.slice.call(arguments, 1));
-    args.unshift(merger);
-    return TypedMap.__coerce(Immutable.Map.prototype.mergeDeepWith.apply(this, args));
-  };
-
-  TypedMap.prototype.__forceIterables = function(args) {
-    // This forces all arguments into an iterable, which makes all the
-    // merge functions avoid calling `fromJS` and converting them
-    // deeply into immutable objects.
-    const iters = [];
-    for(var i=0; i<args.length; i++) {
-      if(!Immutable.Iterable.isIterable(args[i])) {
-        iters.push(Immutable.Iterable(args[i]));
-      }
-      else {
-        iters.push(args[i]);
-      }
-    }
-    return iters;
-  };
+  var Type = t.dict(domain, codomain, name);
 
   function ImDict(value, path) {
-
-    if (process.env.NODE_ENV === 'production') {
-      if (identity) {
-        return (value instanceof Immutable.Map) ? value : Immutable.Map(value);
-      }
+    if (ImDict.is(value)) {
+      return value;
     }
-
-    if (process.env.NODE_ENV !== 'production') {
-      path = path || [displayName];
-      assert(isObject(value) || (value instanceof Immutable.Map), function () { return 'Invalid value ' + assert.stringify(value) + ' supplied to ' + path.join('/'); });
+    if (!(this instanceof ImDict)) {
+      return new ImDict(value, path);
     }
-
-    var idempotent = true;
-    var ret = {};
-    var obj = (value instanceof Immutable.Map) ? value.toObject() : value;
-
-    for (var k in obj) {
-      if (obj.hasOwnProperty(k)) {
-        k = create(domain, k, ( process.env.NODE_ENV !== 'production' ? path.concat(domainNameCache) : null ));
-        var actual = obj[k];
-        var instance = create(codomain, actual, ( process.env.NODE_ENV !== 'production' ? path.concat(k + ': ' + codomainNameCache) : null ));
-        idempotent = idempotent && ( actual === instance );
-        ret[k] = instance;
+    if (process.env.NODE_ENV !== 'production') {
+      if (value instanceof Immutable.Map) {
+        value = value.toObject();
       }
+      return t.mixin(this, Immutable.Map(Type(value, path)), true);
     }
-
-    if (idempotent && value instanceof TypedMap) { // implements idempotency
-      return value;
-    }
-
-    return TypedMap.__coerce(Immutable.Map(ret));
+    return t.mixin(this, Immutable.Map(value), true);
```

Comment:
I see that `mixin` iterates over the object and copies all methods over. I'd be wary of the perf impact in production if using immutable heavily, as every single operation on them will do this. Also, because `mixin` checks `hasOwnProperties`, is there a chance it will miss any inherited methods on immutable objects? (I know Map "inherits" from some base classes, but I don't know if it does it by way of deep prototype chains)